provisioning/transport/amqp/src/amqp.ts

```diff
@@ -314,8 +326,23 @@ export class Amqp extends EventEmitter implements X509ProvisioningTransport, Tpm
             this._senderLink.send(requestMessage, (err) => {
               if (err) {
                 delete this._operations[requestMessage.correlation_id];
-                /*Codes_SRS_NODE_PROVISIONING_AMQP_16_011: [The `registrationRequest` method shall call its callback with an error if the transport fails to send the request message.]*/
-                callback(err);
+                const translatedError = translateError('registration failure', err);
+                /*Codes_SRS_NODE_PROVISIONING_AMQP_06_007: [If the `registrationRequest` send request is rejected with an `InternalError` or `ThrottlingError`, the result.status value will be set with `registering` and the callback will be invoked with *no* error object.] */
+                if ((translatedError instanceof errors.InternalServerError) || ((translatedError as AmqpTransportError) instanceof errors.ThrottlingError)) {
+                  debug('retryable error on registration: ' + err.name);
+                  let retryAfter: number;
+                  /*Codes_SRS_NODE_PROVISIONING_AMQP_06_009: [If the `registrationRequest` rejection error contains the info property`retry-after`, it will be interpreted as the number of seconds that should elapse before the next attempted operation.  Otherwise default.] */
+                  if ((err as any).info && (err as any).info[MessagePropertyNames.retryAfter]) {
+                    retryAfter = Number(((err as any).info[MessagePropertyNames.retryAfter] as string)) * 1000;
+                  } else {
+                    retryAfter = this._config.pollingInterval;
+                  }
```

Comment:
these code sections could be refactored (const retryAfter = retryAfterFromError(err, defaultValue) or const retryAfter = retryAfterFromMessageProperty(err, defaultValue).

Reply:
Indeed, every time I wrote this I thought about it.  In the end it simply always resolved down to an additional line or two of code per section to put a function call to replace an if then else  (although in most cases slightly different enough).  This cure is no improvement.

---
In reply to: [282197924](https://github.com/Azure/azure-iot-sdk-node/pull/555#discussion_r282197924) [](ancestors = 282197924)